src/App.js

```diff
@@ -44,8 +90,164 @@ class App extends Component {
         });
     }
 
+    handleAddCategory(category) {
+        if(category.name === '') return;
+
+        const newCategory = {
+            id: this.state.nextCatId,
+            name: category.name,
+            parentId: null
+        };
+
+        let newList = this.state.categories;
+        newList.unshift(newCategory);
+
+        this.setState({
+            categories: newList,
+            nextCatId: this.state.nextCatId + 1
+        });
```

Comment:
If you need to set the state based on the previous state, you should pass a function to `setState` - for detailed explaination see [React's docs on setState](https://facebook.github.io/react/docs/react-component.html#setstate).

```js
this.setState((prevState) => ({
  categories: newList,
  nextCatId: prevState.nextCatId + 1,
}));
```